### PR TITLE
`azurerm_mssql_server` - enable ability to apply TDE using a CMK at time of server deployment

### LINKED ISSUE #20549: Enable ability to apply azurerm_mssql_server TDE using a CMK at time of server deployment
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

When trying to deploy a MSSQL Server via azurerm_mssql_server and you have Azure policies in place to require a customer-managed key for TDE Terraform currently fails to deploy.  It appears that Terraform will first deploy the SQL server resource in order to get the server_id as required by azurerm_mssql_server_transparent_data_encryption.  However, due to the policy in place, this is denied since the policy is looking for the KeyID on the SQL Server resource which does not exist yet.  It would be great to move this action within the azurerm_mssql_server resource.  If there is a way to do this now please let me know.  Thanks!

### New or Affected Resource(s)/Data Source(s)

azurerm_mssql_server

### Potential Terraform Configuration

```hcl
resource "azurerm_mssql_server" "demo" {
  name        = "demo"
  location    = "eastus"
  version     = "12.0"
  resource_group_name   = "rg-demo"
  azurerm_mssql_server_transparent_data_encryption {
    key_vault_key_id = azurerm_key_vault_key.tde_key.id
  }
  minimum_tls_version = "1.2"
  public_network_access_enabled = "false"
  outbound_network_restriction_enabled = "true"
  primary_user_assigned_identity_id = azurerm_user_assigned_identity.identity.id
  identity {
    type = "UserAssigned"
    identity_ids = [azurerm_user_assigned_identity.identity.id]
  }  
}
```

### References

https://github.com/Azure/azure-policy/blob/master/built-in-policies/policyDefinitions/SQL/SqlServer_EnsureServerTDEisEncryptedWithYourOwnKey_Deny.json

### PR DESCRIPTION
**Pull request by WodansSon**
(fixes #20549)

### CODE REVIEW

**Review by jackofallops — CHANGES_REQUESTED** [review]
Thanks @WodansSon 
Looking mostly good, I've left some comments in-line below if you could take a look?  Also, as discussed offline, the existing TDE resource tests are failing due to the new property and need a lifecycle/ignore_changes config adding to accommodate the dual mechanisms of setting this. 

Thanks!

File: internal/services/mssql/mssql_server_resource.go [hunk 1]
```diff
@@ -245,10 +254,31 @@ func resourceMsSqlServerCreate(d *pluginsdk.ResourceData, meta interface{}) erro
...
+		}
+	}
+
 	if primaryUserAssignedIdentityID, ok := d.GetOk("primary_user_assigned_identity_id"); ok {
 		props.PrimaryUserAssignedIdentityID = utils.String(primaryUserAssignedIdentityID.(string))
 	}
 
+	// if you pass the Key ID you must also define the PrimaryUserAssignedIdentityID...
+	if props.KeyID != nil && props.PrimaryUserAssignedIdentityID == nil {
+		return fmt.Errorf("if the 'key_vault_customer_managed_key_id' field has been defined you must also define the 'primary_user_assigned_identity_id' field as well, got %s", pointer.From(props.KeyID))
```
[the remark this answers is not in the document]
Comment (WodansSon) [reply internal/services/mssql/mssql_server_resource.go:279]: Fixed.

File: internal/services/mssql/mssql_server_resource.go [hunk 2]
```diff
@@ -330,10 +360,30 @@ func resourceMsSqlServerUpdate(d *pluginsdk.ResourceData, meta interface{}) erro
...
 
+	if d.HasChange("key_vault_customer_managed_key_id") {
+		keyVaultKeyId := d.Get(("key_vault_customer_managed_key_id")).(string)
+
+		keyId, err := keyVaultParser.ParseNestedItemID(keyVaultKeyId)
+		if err != nil {
+			return fmt.Errorf("unable to parse key: %q: %+v", keyVaultKeyId, err)
+		}
+
+		if keyId.NestedItemType == "keys" {
```
[the remark this answers is not in the document]
Comment (WodansSon) [reply internal/services/mssql/mssql_server_resource.go:371]: As mentioned above, I will address this in a separate PR.

File: internal/services/mssql/mssql_server_resource.go [hunk 3]
```diff
@@ -330,10 +360,30 @@ func resourceMsSqlServerUpdate(d *pluginsdk.ResourceData, meta interface{}) erro
...
+		}
+	}
+
 	if primaryUserAssignedIdentityID, ok := d.GetOk("primary_user_assigned_identity_id"); ok {
 		props.PrimaryUserAssignedIdentityID = utils.String(primaryUserAssignedIdentityID.(string))
 	}
 
+	// if you pass the Key ID you must also define the PrimaryUserAssignedIdentityID...
+	if props.KeyID != nil && props.PrimaryUserAssignedIdentityID == nil {
+		return fmt.Errorf("if the 'key_vault_customer_managed_key_id' field has been defined you must also define the 'primary_user_assigned_identity_id' field as well, got %s", pointer.From(props.KeyID))
```
[the remark this answers is not in the document]
Comment (WodansSon) [reply internal/services/mssql/mssql_server_resource.go:384]: Fixed.

File: website/docs/r/mssql_server.html.markdown [hunk 4]
```diff
@@ -42,6 +42,86 @@ resource "azurerm_mssql_server" "example" {
...
+    identity_ids = [azurerm_user_assigned_identity.example.id]
+  }
+
+  primary_user_assigned_identity_id = azurerm_user_assigned_identity.example.id
+  key_vault_customer_managed_key_id = azurerm_key_vault_key.example.id
+}
+
+# Create a key vault with access policies which allow for the current user to get, list, create, delete, update, recover, purge and getRotationPolicy for the key vault key and also add a key vault access policy for the Microsoft Sql Server instance User Managed Identity to get, wrap, and unwrap key(s)
+resource "azurerm_key_vault" "example" {
+  name                        = "mssqltdeexampleing"
```
[the remark this answers is not in the document]
Comment (WodansSon) [reply website/docs/r/mssql_server.html.markdown:86]: Fixed.

**Comment by WodansSon** [comment]
Original test pass:
image

Test pass with failed tests now passing:
image

File: internal/services/mssql/mssql_server_resource.go [hunk 5]
```diff
@@ -245,10 +254,31 @@ func resourceMsSqlServerCreate(d *pluginsdk.ResourceData, meta interface{}) erro
...
 
+	if v, ok := d.GetOk("key_vault_customer_managed_key_id"); ok {
+		keyVaultKeyId := v.(string)
+
+		keyId, err := keyVaultParser.ParseNestedItemID(keyVaultKeyId)
+		if err != nil {
+			return fmt.Errorf("unable to parse key: %q: %+v", keyVaultKeyId, err)
+		}
+
+		if keyId.NestedItemType == "keys" {
```
[the remark this answers is not in the document]
Comment (WodansSon) [reply internal/services/mssql/mssql_server_resource.go:265]: @jackofallops: I have opened PR #21767 to address this comment.

**Review by katbyte — APPROVED** [review]
LGTM 🔑

**Pull request merged**